python/task_manager/results_poller.py

```diff
@@ -9,103 +9,182 @@
 # not expressly granted therein are reserved by Shotgun Software Inc.
 
 """
-Results queue poller
+Results dispatcher for the background task manager.
 """
 
 import Queue
 from sgtk.platform.qt import QtCore
 import sgtk
 
 
-class ResultsPoller(QtCore.QObject):
+class _TaskCompletedEvent(object):
     """
-    Polls a queue which holds the results posted from the worker threads.
+    Event sent when a task is succesfully completed.
     """
 
+    def __init__(self, worker_thread, task, result):
+        """
+        Constructor.
+
+        :param worker_thread: Worker thread that completed sucessfully.
+        :param task: Task id of the completed task.
+        :param result: Result published by the task.
+        """
+        self.worker_thread = worker_thread
+        self.task = task
+        self.result = result
+
+
+class _TaskFailedEvent(object):
+    """
+    Event sent when a task is succesfully completed.
+    """
+
+    def __init__(self, worker_thread, task, message, traceback):
+        """
+        Constructor.
+
+        :param worker_thread: Worker thread that completed sucessfully.
+        :param task: Task id of the completed task.
+        :param message: Error message from the worker thread.
+        :param traceback: Traceback from  the worker thread error.
+        """
+        self.worker_thread = worker_thread
+        self.task = task
+        self.message = message
+        self.traceback = traceback
+
+
+class ResultsDispatcher(QtCore.QThread):
+    """
+    Dispatches events synchronously to the thread that owns this object.
+
+    Signalling between two different threads in PySide is broken in several versions
+    of PySide. There are very subtle race conditions that arise when there is a lot
+    of signalling between two threads. Some of these things have been fixed in later
+    versions of PySide, but most hosts integrate PySide 1.2.2 and lower, which are
+    victim of this race condition.
+
+    The background task manager does a lot on inter-threads communications and
+    therefore can easily fall pray to these deadlocks that exist within PySide.
+
+    Therefore, we instead use Qt's QMetaObject invokeMethod to carry information
+    to the background task manager thread in a thread-safe manner, since it
+    doesn't exhibit the bad behaviour from PySide's signals.
+    """
+
+    class _ShutdownHint(object):
```

Comment:
With the latest changes, should this be moved out of the class and be named `_ShutdownRequestedEvent` for consistency with `_TaskFailedEvent` and `_TaskCompletedEvent`?